List Operations (#57)
exercism/awk: exercises/practice/list-ops/.docs/instructions.append.md

# placeholder

## New AWK concepts

This is the first exercise where the solution is not a "main" script.
The goal is to write a library of functions to be included into other awk programs.
There are several interesting new concepts to see in this exercise.

The following sections that are tagged with "(gawk)" are specifically GNU awk extensions.

### Include files <sup>(gawk)</sup>

Notice the [`@include`][d-include] directive in the test
file.  This instructs gawk to read and evaluate the named file.

### Namespaces <sup>(gawk)</sup>

> All awk variables are global

Reviewer: The downside of fixed-width text :) If you make edits, you need to reformat. Or live with unbalanced blocks -- which I'm totally fine with.

Author: I know. This one was in-between: I intended to make the review easier

### Dynamic function invocation <sup>(gawk)</sup>

awk functions are not first-class objects: they can't get passed around like
some other languages allow. However, the _name_ of a function can be passed as a
string and that is used to invoke the function, using a special
`@varname(args)` notation. An example:

```awk
function greet(name) {
    print "Hello, " name
}
BEGIN {
    greet("John")       # => "Hello, John"

    f = "greet"
    f("Bill")           # => error
    @f("Bill")          # => "Hello, Bill"
}
```
This is described in [Indirect Function Calls][indirect].

### How function parameters are passed

Array parameters are passed **by reference**. Changes to the array made in
the function are visible in the caller.

Non-array parameters are passed **by value**.

For _untyped_ parameters, it depends on what the function does with them:
- if the function initializes it as an array, then it becomes a parameter
  passed by reference,
- if the function initializes it as a scalar value (a number or a string),
  then it is not a reference.

Full details are in the manual in [Passing Function Arguments by Value Or by
Reference][pass-by].

#### Local variables

> "Pass by value" parameters is how to achieve local variables in a
> function. It is not an error to pass fewer values to a function than the

Reviewer: ```suggestion
Function scoped (local) variables can be created by using "pass by value" parameters. It is not an error to pass fewer values to a function than the
```

Author: Oops, merged too soon. I'll add this

# placeholder

## New AWK concepts

There are several interesting new concepts we're seeing in this exercise.

### Include files

> This is the first exercise we've seen where the solution we're writing is

Reviewer: In general, I try to avoid "we" or "you" in tech docs. Talk about the exercise and not the author.

not a "main" script. We're writing a library to be included into other awk
programs.  You'll see the [`@include`][d-include] directive in the test

> file.  This is analagous to the shell's `source` (aka `.`) command.

Reviewer: If you're going to describe what `@include` does, consider describing it without just referencing another language.

This is a GNU awk extension.

### Namespaces

> We're introducing [namespaces][namespaces]. All awk variables are global

Reviewer: ```suggestion
This exercise introduces [namespaces][namespaces]. All awk variables are global
```

> (except for function parameters which are local to the function).  Clearly

Reviewer: "Clearly" suggests the reader ought to have come to some conclusion on their own and is slow if they failed to do so.
